application_handlers_test.go

```diff
@@ -263,7 +318,35 @@ func TestApplicationListBadRequestInvalidFilter(t *testing.T) {
 		t.Error(err)
 	}
 
-	testutils.BadRequestTest(t, rec)
+	templates.BadRequestTest(t, rec)
+}
+
+func TestApplicationListWithOffsetAndLimit(t *testing.T) {
+	helpers.SkipIfNotRunningIntegrationTests(t)
+	testData := templates.TestDataForOffsetLimitTest
+	wantApplicationsCount := len(fixtures.TestApplicationData)
+
+	for _, i := range testData {
+		c, rec := request.CreateTestContext(
+			http.MethodGet,
+			"/api/sources/v3.1/applications",
+			nil,
+			map[string]interface{}{
+				"limit":    i["limit"],
+				"offset":   i["offset"],
+				"filters":  []util.Filter{},
+				"tenantID": int64(1),
+			},
+		)
+
+		err := ApplicationList(c)
+		if err != nil {
+			t.Error(err)
+		}
+
+		path := c.Request().RequestURI
+		templates.WithOffsetAndLimitTest(t, path, rec, wantApplicationsCount, i["limit"], i["offset"])
+	}
 }
```

Comment:
If I recall correctly, the bug affected the DAO functions which had an offset different than zero. That is the reason why, in my opinion, it is not needed to test the handlers themselves but just the affected DAO functions. I was thinking on something like this for the `List*` functions:

```go
// TestApplicationListOffsetRegression is a regression test for RHCLOUD-18515. The test tests that the offset doesn't
// affect the returned total count.
func TestApplicationListOffsetRegression(t *testing.T) {
	testutils.SkipIfNotRunningIntegrationTests(t)
	CreateFixtures("applicationoffsetregression")

	applicationDao := GetApplicationDao(&fixtures.TestTenantData[0].Id)

	// create a new source and some applications in a loop.

	offset := 5
	_, count, err := applicationDao.List(0, offset, []util.Filter{})
	if err != nil {
		t.Errorf(`unexpected error when listing the applications: %s`, err)
	}

	{
		want := // total count of the created fixtures.
		got := count

		if want != got {
			t.Errorf(`incorrect total count returned. Want "%d", got "%d"`, want, got)
		}
	}
}
```

With this we ensure that we won't get hit by a regression in the future, and the test is focused on the bug itself exactly where it was located.

Reply:
yes, that makes sense ... I didn't realise that is not needed to go from handlers in this case